### PR TITLE
Improve gp_toolkit.gp_check_orphaned_files for more reliable results

### PR DESCRIPTION
**Pull request by huansong**
This commit mainly improves the gp_check_orphaned_files view in the sense that, since ultimately its user is likely going to remove the reported orphaned files, we would not like that to cause any potential issues due to causes like:

* Main relation files that are associated with dropped tables are kept until the next CHECKPOINT in order to prevent potential issue with crash recovery (see comments for mdunlink()). So removing them would have issue.
* Relation files that are created during an ongoing transactions could be recognized as orphaned: another session will see an old pg_class.relfilenode so it would think the new relfilenode is orphaned. So if one removes that, we might have data loss.

So accordingly, the improvements are:
* We should force a CHECKPOINT prior to collecting the orphaned file list.
* We should exclude other activities that might change pg_class.relfilenode while running the view. This is done by locking pg_class in SHARE mode (which blocks writes but allows read) with "nowait" flag (which allows the lock attempt to immediately return so we are not blocked forever). We should also check pg_stat_activity to make sure that there is no idle transaction (because the idle transaction might've already modified pg_class and released the lock). In the new view we do that by simply making sure there's no concurrent client sessions. This might be a bit too restricted but it's better to be as safe as we can in sensitive area like this.

These steps will need to be written in a function. So rewrite the gp_check_orphaned_files view to be SELECT'ing from a new UDF.

Also improve the view results by adding relative path of each file being reported for convenience about further action of the files.

Dev pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/orphaned-view-improve

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpcontrib/gp_toolkit/gp_toolkit--1.1--1.2.sql [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    BEGIN
+        -- lock pg_class so that no one will be adding/altering relfilenodes
+        LOCK TABLE pg_class IN SHARE MODE NOWAIT;
+       
+        -- make sure no other active/idle transaction is running
+        IF EXISTS (
+            SELECT 1
+            FROM gp_stat_activity
+            WHERE
+            sess_id <> -1 AND backend_type IN ('client backend', 'unknown process type') -- execlude background worker types
```
Comment (bmdoil) [inline gpcontrib/gp_toolkit/gp_toolkit--1.1--1.2.sql:57]: nit: s/execlude/exclude

File: gpcontrib/gp_toolkit/gp_toolkit--1.1--1.2.sql [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            AND sess_id <> current_setting('gp_session_id')::int -- Exclude the current session
+        ) THEN
+            RAISE EXCEPTION 'There is client session running on one or more segment. Aborting...';
```
Comment (bmdoil) [inline gpcontrib/gp_toolkit/gp_toolkit--1.1--1.2.sql:60]: nit: `There is a client session running on one or more segments. Aborting...`

File: src/test/regress/input/gp_check_files.source [hunk 3]
```diff
@@ -29,6 +26,17 @@ select get_tablespace_version_directory_name() as version_dir \gset
...
+\! touch 987654
+\! touch 987654.3
+
+-- check orphaned files, note that this forces a checkpoint internally.
+select gp_segment_id, filename from gp_toolkit.gp_check_orphaned_files;
+
+-- remove the orphaned files so not affect subsequent tests
+\! rm 987654
+\! rm 987654.3
+
```
Comment (bmdoil) [inline src/test/regress/input/gp_check_files.source:39]: Is it possible to move this test to `gpcontrib/gp_toolkit/sql`?

File: src/test/isolation2/sql/misc.sql [hunk 4]
```diff
@@ -53,3 +53,16 @@ create or replace view misc_v as select 1;
...
+--
+-- use a different database to do the test, otherwise we might be reporting tons 
+-- of orphaned files produced by the many intential PANICs/restarts in the isolation2 tests.
+create database check_orphaned_db;
+1:@db_name check_orphaned_db: begin;
+2:@db_name check_orphaned_db: select * from gp_toolkit.gp_check_orphaned_files;
+1q:
+2q:
+
+drop database check_orphaned_db;
```
Comment (bmdoil) [inline src/test/isolation2/sql/misc.sql:68]: Is it possible to move this test to `gpcontrib/gp_toolkit/sql` or do we need the isolation2 framework?

File: src/test/regress/input/gp_check_files.source (hunk 3 above)
Comment (huansong) [reply src/test/regress/input/gp_check_files.source:39]: Technically possible. Though historically all gp_toolkit tests were running under regress or isolation2 tests so I was not looking to move stuff here unless we have made that decision. Plus, another purpose of this test was intended to catch problems caused by the regress tests (we have caught quite a few). So running this in `regress` makes slightly more sense (though nothing prevents us from just connecting to the `regression` database in `gpcontrib/gp_toolkit` test, just a bit weird).

P.S. I just found that `gp_contrib/gp_toolkit` tests were not actually being run in CI... Fixed that in another commit.

File: src/test/isolation2/sql/misc.sql (hunk 4 above)
Comment (huansong) [reply src/test/isolation2/sql/misc.sql:68]: We need the isolation2 framework for this test snipet.

File: src/test/regress/input/gp_check_files.source (hunk 3 above)
Comment (bmdoil) [reply src/test/regress/input/gp_check_files.source:39]: Makes sense, thanks!

**Review by bmdoil — APPROVED** [review]
LGTM

nit: s/Improve/Improve

Is this something that might benefit upstream as well?

**Comment by huansong** [comment]
> Is this something that might benefit upstream as well?

It's good as an external tool but I don't think it'll make into the upstream repo based on comments like: https://www.postgresql.org/message-id/20170816171551.moi5hzgomtmvql6q%40alap3.anarazel.de . Real core feature like https://github.com/greenplum-db/gpdb/pull/16325 is more likely.

Btw, I fixed a flakiness w/ the test, where background worker could make `LOCK pg_class` failed randomly - now we just repeatedly try `gp_toolkit.gp_check_orphaned_files` in 10 minutes which should be well enough to avoid that flakiness.

**Pull request merged**